source/agora/stats/Stats.d

```diff
@@ -0,0 +1,163 @@
+/*******************************************************************************
+
+    Base stats class that helps define other stats classes
+
+    Copyright:
+        Copyright (c) 2019-2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.stats.Stats;
+
+import agora.utils.Utility : underscoreSeparatedToUpperCamelCase;
+
+import std.array : join, split, byPair;
+import std.typecons : Tuple, tuple;
+
+/***************************************************************************
+
+        Base stats class that helps define other stats classes.class
+        Usage is Stats!(ConcreateStatsValue, ConcreateStatsLabel) concreate_stats;
+
+        Params:
+            ValueType = fields of this struct will correspond to metric names
+            LabelType = fields of this struct will correspond to label names
+
+***************************************************************************/
+
+public struct Stats (ValueType, LabelType)
+{
+
+    public alias ValueType ValueTypeT;
+    public alias LabelType LabelTypeT;
+    public alias StatValueLabelTup = Tuple!(const ValueType, "value", const LabelType, "label");
+    private ValueType[string] stats_maps;
+
+    private static immutable SEPARATOR = "|SEPARATORSYM|";
+
+    /// generating functions with signature
+    /// increase<CamelCaseNameOfMetric>By
+    ///    (LabelTs...)(ulong number, LabelTs labels_packed) pure nothrow @safe
+    static foreach (member_name; __traits(allMembers, ValueType))
+    {
```

Comment:
I am going to change all the mixins to use the q{...} tmr, but let me know any other problems

Reply:
Wouldn't it be possible to instead write it to something like:

```D
public void increaseStats (string name, LabelTs...)(ulong number, LabelTs labels)
```

That way we can actually look up functions. Right now with the use of mixins we cannot go to the function definition easily.

Also we should avoid using mixins here, we can use `__traits(getMember, "name")` for this use-case.

Reply:
> Wouldn't it be possible to instead write it to something like:
> 
> ```d
> public void increaseStats (string name, LabelTs...)(ulong number, LabelTs labels)
> ```
> 
> That way we can actually look up functions. Right now with the use of mixins we cannot go to the function definition easily.
> 
> Also we should avoid using mixins here, we can use `__traits(getMember, object, "name")` for this use-case.

that is a good idea, we can get rid of the mixins that way